lib/ace/mode/haml_highlight_rules.js

```diff
@@ -11,116 +11,130 @@ var HamlHighlightRules = function() {
     // regexp must not have capturing parentheses. Use (?:) instead.
     // regexps are ordered -> the first match is used
     HtmlHighlightRules.call(this);
-    this.$rules.start.unshift(
-        {
-            token : "punctuation.section.comment",
-            regex : /^\s*\/.*/
-        },
-        {
-            token: "string.quoted.double",
-            regex: "==.+?=="
-        },
-        {
-            token: "keyword.other.doctype",
-            regex: "^!!!\\s*(?:[a-zA-Z0-9-_]+)?"
-        },
-        RubyExports.qString,
-        RubyExports.qqString,
-        RubyExports.tString,
-        {
-            token: "character.escape.haml",
-            regex: "^\\s*\\\\."
-        },
-        {
-            token: "text",
-            regex: /^\s*/,
-            next: "tag_single"
-        },
-        RubyExports.constantNumericHex,
-        RubyExports.constantNumericFloat,
-        
-        RubyExports.constantOtherSymbol,
-        {
-            token: "text",
-            regex: "=|-|~",
-            next: "embedded_ruby"
-        }
-    );
-    this.$rules.tag_single = [
-        {
-            token: "meta.tag.haml",
-            regex: /(%[\w:\-]+)/
-        },
-        {
-            token: "keyword.attribute-name.class.haml",
-            regex: "\\.[\\w-]+"
-        },
-        {
-            token: "keyword.attribute-name.id.haml",
-            regex: "#[\\w-]+"
-        },
-        {
-            token: "punctuation.section",
-            regex: "\\{",
-            next: "section"
-        },
-        
-        RubyExports.constantOtherSymbol,
-        
-        {
-            token: "text",
-            regex: /\s/,
-            next: "start"
-        },
-        {
-            token: "empty",
-            regex: "$|(?!\\.|#|\\{|\\[|=|-|~|\\/)",
-            next: "start"
-        }
-    ];
-    this.$rules.section = [
-        RubyExports.constantOtherSymbol,
-        
-        RubyExports.qString,
-        RubyExports.qqString,
-        RubyExports.tString,
-        
-        RubyExports.constantNumericHex,
-        RubyExports.constantNumericFloat,
-        {
-            token: "punctuation.section",
-            regex: "\\}",
-            next: "start"
-        } 
-    ];
-    
-    this.$rules.embedded_ruby = [ 
-        RubyExports.constantNumericHex,
-        RubyExports.constantNumericFloat,
-        {
+
+    this.$rules = {
+        "start": [
+            {
+                token: "comment.block", // multiline HTML comment
+                regex: /^\/$/,
+                next: "comment"
+            },
+            {
+                token: "comment.block", // multiline HAML comment
+                regex: /^\-#$/,
+                next: "comment"
+            },
+            {
+                token: "comment.line", // HTML comment
+                regex: /\/\s*.*/
+            },
+            {
+                token: "comment.line", // HAML comment
+                regex: /-#\s*.*/
```

Comment:
I would say both, based on the [HAML reference](http://haml.info/docs/yardoc/file.REFERENCE.html#comments). I was not sure if Ace already had a way of detecting indentation, so I rolled out a custom a approach, which proved to be a bit difficult but covers most cases.

According to the reference, comments can be single line or multiline, and multiline comments are implemented using indentation. So I added support for both.

I took a look at the YAML indentation implementation. Looks nice. Is this the way indentation detected should be implemented for Ace?